modules/blocks/index.js

```diff
@@ -64,3 +71,56 @@ export function getBlockSettings( slug ) {
 export function getBlocks() {
 	return Object.values( blocks );
 }
+
+/**
+ * Iterates through children of a given element to discover block nodes.
+ *
+ * @param  {Element}  rootNode Element to iterate
+ * @return {Object[]}          Block nodes
+ */
+export function findBlockNodes( rootNode ) {
+	const nodes = [];
+
+	let child = rootNode.firstChild;
+	while ( child ) {
+		// Test as opening comment
+		if ( Node.COMMENT_NODE !== child.nodeType || ! RE_BLOCK_OPEN.test( child.nodeValue ) ) {
+			child = child.nextSibling;
+			continue;
+		}
+
+		// Found opening block comment
+		const match = child.nodeValue.match( RE_BLOCK_OPEN );
+		const closeTag = '/' + child.nodeValue.trim();
+		const block = {
+			type: match[ 1 ],
+			startNode: child,
+			blockNodes: []
+		};
```

Comment:
@joyously I think your suggestion will work fine for new dynamic blocks if we choose to leverage shortcodes for those, but in the case of a static block (text, images, etc) the hope is to represent those in `post_content` as HTML for reasons identified in the [Editor Technical Overview](https://make.wordpress.org/core/2017/01/17/editor-technical-overview/) post.

So in the case of an quote block, we'd want to identify both the text of the quote and an optional citation from the HTML in `post_content`. There's been some discussion in #104 about options we could explore to achieve this. The point I'd made in https://github.com/WordPress/gutenberg/pull/306#discussion_r107527146 is that you could use DOM APIs like [`querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) to get at the respective markup nodes (`<p>`, `<cite>`).